src/elf/gnu_hash.rs

```diff
@@ -1,143 +1,210 @@
+/// used to compute a bucket index.
+///
+/// Consequently, if the hashing function returns the value `x` for some name,
+/// `buckets[x % nbuckets]` gives an index, `y`, into both the symbol table
+/// and the chain table.
 pub fn hash(symbol: &str) -> u32 {
     const HASH_SEED: u32 = 5381;
-    let mut hash = HASH_SEED;
-    for b in symbol.as_bytes() {
-        hash = hash
-            .wrapping_mul(33)
-            .wrapping_add(u32::from(*b));
-    }
-    hash
+    symbol
+        .bytes()
+        .fold(HASH_SEED, |hash, b| hash.wrapping_mul(33).wrapping_add(u32::from(b)))
 }
 
 #[cfg(test)]
 mod tests {
     use super::hash;
     #[test]
     fn test_hash() {
-        assert_eq!(hash("")             , 0x0000_1505);
-        assert_eq!(hash("printf")       , 0x156b_2bb8);
-        assert_eq!(hash("exit")         , 0x7c96_7e3f);
-        assert_eq!(hash("syscall")      , 0xbac2_12a0);
+        assert_eq!(hash(""), 0x0000_1505);
+        assert_eq!(hash("printf"), 0x156b_2bb8);
+        assert_eq!(hash("exit"), 0x7c96_7e3f);
+        assert_eq!(hash("syscall"), 0xbac2_12a0);
         assert_eq!(hash("flapenguin.me"), 0x8ae9_f18e);
     }
 }
 
 macro_rules! elf_gnu_hash_impl {
-    ($size:ty) => {
-
-        use core::slice;
-        use core::mem;
+    ($IntTy:ty) => {
+        use crate::elf::sym::Sym;
         use crate::strtab::Strtab;
-        use super::sym;
+        use core::fmt;
+        use core::mem;
+        use core::slice;
+
+        const INT_SIZE: usize = mem::size_of::<$IntTy>();
+        const U32_SIZE: usize = mem::size_of::<u32>();
+        /// Size of a bits mask in bloom filter
+        const ELFCLASS_BITS: u32 = INT_SIZE as u32 * 8;
 
-        pub struct GnuHash<'process> {
-            nbuckets: u32,
-            symindex: usize,
+        /// A better hash table for the ELF used by GNU systems in GNU-compatible software.
+        pub struct GnuHash<'a> {
+            /// Index of the first symbol in the `.dynsym` table which is accessible with
+            /// the hash table
+            symindex: u32,
+            /// Shift count used in the bloom filter
             shift2: u32,
-            maskbits: u32,
-            bloomwords: &'process [$size], // either 32 or 64 bit masks, depending on platform
-            maskwords_bitmask: u32,
-            buckets: &'process [u32],
-            hashvalues: &'process [u32],
-            symtab: &'process [sym::Sym],
+            /// 2 bit bloom filter on `chains`
+            // Either 32 or 64-bit depending on the class of object
+            bloom_filter: &'a [$IntTy],
+            /// GNU hash table bucket array; indexes start at 0. This array holds symbol
+            /// table indexes and contains the index of hashes in `chains`
+            buckets: &'a [u32],
+            /// Hash values; indexes start at 0. This array holds symbol table indexes.
+            chains: &'a [u32], // => chains[total_dynsyms - symindex]
```

Comment:
I should find a way to encode the dynsyms len.
Or should I save dynsyms in this struct?

Reply:
So you moved `dynsyms` into this struct. Why? I don't really care if it is in this struct or not (@m4b might though), but I would like to understand what the problem is that you want to solve by doing this.

Reply:
The `dynsyms` was in this struct before this PR. I moved it out because I want
to use `impl IntoIterator<Sym>` which you pointed out it may be slow in some
situations when combined with `Iterator::skip` method.
I move it in again because now we instead of passing `total_dynsyms` (which is `dynsyms.len()`)
to construct `chains` array. Just use `dynsyms` instead.

Reply:
I see, thanks!